examples/platforms/simulation/infra_if.c

```diff
@@ -36,7 +36,7 @@
 #include "simul_utils.h"
 #include "utils/code_utils.h"
 
-#if OPENTHREAD_CONFIG_BORDER_ROUTING_ENABLE && !OPENTHREAD_RADIO
+#if OPENTHREAD_SIMULATION_IMPLEMENT_INFRA_IF
```

Comment:
What is the reason for adding this?

Reply:
The intention for adding this is to implement a NCP version of `otPlatInfraIf` APIs. The NCP version implementation would be different as the simulation version. For example, `otPlatInfraIfSendIcmp6Nd` will be implemented as NCP sending a message to the host through spinel, and the host sends the Icmp6 Nd message.

The simulation impl will then conflict with the NCP impl. So this PR tries to add a specific switch for the simulation infra if to avoid conflict when we want to use the NCP version. Thoughts?

Reply:
Should we consider an NCP-based implementation its own platform to distinguish it from "simulation platform"?

Reply:
I actually considered about that. But it seems that most platform APIs don't need an NCP-based implementation and still depend on the platform. Only a few APIs need NCP-based implementation: `otPlatInfraIf`, `otPlatDnssd`. So if we make NCP as a seperate platform, it will miss the implementation for those platform APIs. For example, `flash`, `radio`, `crypto`, etc. We will still need to add implementation for this APIs into NCP platform. 

That means, if we build a simulation node, we need a `NCP-simulation` platform which has NCP-based implementation for `otPlatInfraIf`, `otPlatDnssd` and simulation implementation for other platform APIs. If we build a nordic node, we need another `NCP-nrf` platform. I think that will bring too many platforms and redundant code. So my preference is that we still treat the platforms as they are. And if we use flag control to replace the implementation of `otPlatInfraIf` and `otPlatDnssd` to NCP's implemenation if we want. Thoughts?

Reply:
So your idea is to add this to allow a simulation platform that acts as an NCP (for testing purposes)?

Overall, that sounds reasonable to me. However, I'd suggest we add this flag when we have an actual NCP implementation for `infra-if`. 

Currently, this PR adds `OT_SIMULATION_INFRA_IF`, which simply removes all the `otPlatInfraIf` implementations under simulation without adding a replacement. It might make more sense to introduce a config option when we add a new implementation (like the NCP-based one), allowing us to select which version to use. 

Thoughts?

Reply:
> So your idea is to add this to allow a simulation platform that acts as an NCP (for testing purposes)?
Yes!

Make sense! I'll add the flag together with the new implementation.

Reply:
@abtink 

I've updated to a new approach which I think is better. It's made into a multi option similar as `OT_LOG_OUTPUT`. We can choose `APP` or `PLATFORM_DEFINED`. If we want the NCP based implementation, we choose `APP`. 

I also added an empty NCP based implementation. I would like to first align on this framework and then add implementation gradually with tests.

Reply:
Thanks, @Irving-cl  for adding this. 

While I like the overall approach, I think such a configuration might be better suited within a specific platform, for example under POSIX `OPENTHREAD_POSIX_CONFIG_{}` or within the simulation platform, rather than being part of the main OT core configs. 

I understand you're following the `OPENTHREAD_CONFIG_LOG_OUTPUT` model, but to be honest, I've always felt that `LOG_OUTPUT` was somewhat out of place. However, since it's already there, we're stuck with it. I don't think it's a good model for us to replicate.

My recommendation is to keep `OPENTHREAD_CONFIG` configs limited to those that directly configure something within the OT core itself. In this case, the OT Core doesn't really care where the `otPlatInfraIf` functions are defined. It's really up to any specific platform, if they have multiple implementations, to specify which one to use and/or delegate that implemenation to other modules. 

Thoughts?

Reply:
That makes sense to me!

I plan to put the NCP based implementation under `src/ncp/platform` (a new directory). And for any platform(simulation, nordic, silabs), the implemenation for these APIs (`otPlatInfraIf`) will use the same NCP based implementation. However `simulation` is a special one - it has two implementations. And I will still use the control flag (earlier in this PR) so that we can choose. And the flag will be within the simulation config.

Reply:
@abtink Updated! Thanks.